_user/rotrectangle/main.c

```diff
@@ -0,0 +1,277 @@
+			getchar();
+			ch = getchar();
+		}
+		switch (ch) {
+			case 67:
+				if (self->rec.xc < (self->maxxc - 4))
+					self->rec.xc += 5;
+				break;
+			case 68:
+				if (self->rec.xc > (self->minxc + 4))
+					self->rec.xc -= 5;
+				break;
+			case 65:
+				if (self->rec.yc > (self->minyc + 4))
+					self->rec.yc -= 5;
+				break;
+			case 66:
+				if (self->rec.yc < (self->maxyc - 4))
+					self->rec.yc += 5;
+				break;
+			case 43:
+				if (self->speed < 5)
+					self->speed++;
+				break;
+			case 45:
+				if (self->speed > 1)
+					self->speed--;
+				break;
+			case 'q':
+				self->stop = 1;
+				break;
+			default:
+				break;
+		}
+	}
+
+	endthread();
+}
+
+
+static int rotrectangle_print(rectangle_t *self, graph_t *graphp, unsigned int angle)
+{
+	double dx, dy, alfa, xc, yc, a, b;
+	unsigned int x0, y0, x, y; /* (x,y) is start point for next rectangle lines */
+
+	xc = (double)self->xc;
+	yc = (double)self->yc;
+	a = (double)self->a;
+	b = (double)self->b;
+
+	alfa = ((double)angle * M_PI) / 180.0;
+	x0 = (unsigned int)(xc - ((sin(M_PI / 2 - alfa - atan(b / a))) * sqrt(pow(a, 2.0) + pow(b, 2.0))) / 2.0);
+	y0 = (unsigned int)(yc - ((cos(M_PI / 2 - alfa - atan(b / a))) * sqrt(pow(a, 2.0) + pow(b, 2.0))) / 2.0);
+
+	x = x0;
+	y = y0;
+	dx = (self->a) * cos(alfa);
+	dy = (self->a) * sin(alfa);
+	graph_line(graphp, x, y, (int)dx, (int)dy, 1, self->color, GRAPH_QUEUE_HIGH);
+
+	x = x0 + dx;
+	y = y0 + dy;
+	dx = (self->b) * cos(alfa + (0.5 * M_PI));
+	dy = (self->b) * sin(alfa + (0.5 * M_PI));
+	graph_line(graphp, x, y, (int)dx, (int)dy, 1, self->color, GRAPH_QUEUE_HIGH);
+
+	x = x0;
+	y = y0;
+	dx = (self->b) * cos(alfa + (0.5 * M_PI));
+	dy = (self->b) * sin(alfa + (0.5 * M_PI));
+	graph_line(graphp, x, y, (int)dx, (int)dy, 1, self->color, GRAPH_QUEUE_HIGH);
+
+	x = x0 + dx;
+	y = y0 + dy;
+	dx = (self->a) * cos(alfa);
+	dy = (self->a) * sin(alfa);
+	graph_line(graphp, x, y, (int)dx, (int)dy, 1, self->color, GRAPH_QUEUE_HIGH);
+
+	return 0;
+}
+
+
+static int rotrectangle_save(rotrectangle_t *self)
+{
+	self->prev.xc = self->rec.xc;
+	self->prev.yc = self->rec.yc;
+	self->prev.a = self->rec.a;
+	self->prev.b = self->rec.b;
+	self->prev.color = 0;
+
+	return 0;
+}
+
+
+static int rotrectangle_rotating(rotrectangle_t *self, graph_t *graphp)
+{
+	unsigned int alfa, ret;
+	ret = 0;
+	while (self->stop == 0) {
+		for (alfa = 0; alfa < 360; alfa += self->speed) {
+			if (self->stop != 0)
+				break;
+			ret = rotrectangle_print(&(self->rec), graphp, alfa);
+			/* filling rectangle functions are commented out, because of problem with screen refreshing, it's demo without it */
+			/* graph_fill(graphp, self->rec.xc, self->rec.yc, self->rec.color, GRAPH_FILL_FLOOD, GRAPH_QUEUE_HIGH); */
+			ret = rotrectangle_save(self);
+			usleep(10000); /* min value */
+			ret = rotrectangle_print(&(self->prev), graphp, alfa);
+			/* graph_fill(graphp, self->rec.xc, self->rec.yc, 0, GRAPH_FILL_FLOOD, GRAPH_QUEUE_HIGH); */
+		}
+	}
+
+	return ret;
```

Comment:
not sure why `ret` is needed as it's never checked and all of the functions return 0